sfge/src/main/java/com/salesforce/rules/ops/boundary/SharingPolicyBoundary.java

```diff
@@ -0,0 +1,130 @@
+     */
+    private UserClassVertex applicableAncestorVertex;
+
+    /**
+     * store the {@link UserClassVertex} associated with this boundary.
+     *
+     * @param boundaryItem the {@link UserClassVertex} that the boundary represents
+     * @param g a {@link GraphTraversalSource} used to find the applicable ancestor. See {@link
+     *     #getApplicableAncestorPolicyVertex()}.
+     */
+    public SharingPolicyBoundary(UserClassVertex boundaryItem, GraphTraversalSource g) {
+        this.boundaryItem = boundaryItem;
+        this.g = g;
+        // save the work of recursively querying the graph until it's required
+        this.applicableAncestorVertex = null;
+    }
+
+    /**
+     * @return the {@link UserClassVertex} boundary item
+     */
+    @Override
+    public UserClassVertex getBoundaryItem() {
+        return this.boundaryItem;
+    }
+
+    /**
+     * Try to find the applicable ancestor to the boundary item. The applicable ancestor is defined
+     * as follows:
+     *
+     * <ul>
+     *   <li>If the boundary item has an explicitly defined sharing policy, the applicable ancestor
+     *       will be the boundary item.
+     *   <li>If the boundary item has no explicitly defined sharing policy, the applicable ancestor
+     *       will be the nearest ancestor with an explicitly defined sharing policy.
+     *   <li>If the boundary item has no explicitly defined sharing policy and no ancestors that
+     *       have an explicitly defined sharing policy, the applicable ancestor will be the
+     *       supermost {@link UserClassVertex} ancestor.
+     *   <li>If the boundary item has no explicitly defined sharing policy and no ancestors, the
+     *       applicable ancestor will be the boundary item.
+     *
+     * @return a {@link UserClassVertex} that is the applicable ancestor of the boundary item.
+     */
+    public UserClassVertex getApplicableAncestorPolicyVertex() {
+        // save the work of querying the graph until it's required
+        if (this.applicableAncestorVertex == null) {
+            this.applicableAncestorVertex =
+                    findAncestorWithExplicitSharingPolicy(this.boundaryItem).orElse(boundaryItem);
+        }
+        return this.applicableAncestorVertex;
+    }
+
+    /**
+     * Attempt to find an applicable ancestor of <code>child</code> that has an explicitly defined
+     * sharing policy (with, without, or inherited sharing). See {@link
+     * #getApplicableAncestorPolicyVertex()}
+     *
+     * @param child a {@link UserClassVertex} for which we are attempting to find an ancestor with
+     *     an explicitly defined sharing policy
+     * @return An optional of a {@link UserClassVertex} if an ancestor with an explicitly defined
+     *     sharing policy is found. Empty if there are no such ancestors (or no ancestors at all)
+     */
+    private Optional<UserClassVertex> findAncestorWithExplicitSharingPolicy(UserClassVertex child) {
```

Comment:
If this is only ever invoked from `getApplicableAncestorPolicyVertex`, then you don't need the parameter. You can just use `this.boundaryItem`.